distributed/pytest_resourceleaks.py

```diff
@@ -1,313 +1,365 @@
-        if unknown:
-            raise ValueError(f"unknown resources: {unknown!r}")
-
-        checkers = [all_checkers[leak]() for leak in leaks]
-        checker = LeakChecker(
-            checkers=checkers,
-            grace_delay=config.getvalue("leaks_timeout"),
-            mark_failed=config.getvalue("leaks_mark_failed"),
-            max_retries=config.getvalue("leak_retries"),
-        )
-        config.pluginmanager.register(checker, "leaks_checker")
+    if not leaks:
+        return
+    if leaks == "all":
+        leaks = sorted(c for c in all_checkers if c != "demo")
+    else:
+        leaks = leaks.split(",")
+    unknown = sorted(set(leaks) - set(all_checkers))
+    if unknown:
+        raise ValueError(f"unknown resources: {unknown!r}")
+
+    checkers = [all_checkers[leak]() for leak in leaks]
+    checker = LeakChecker(
+        checkers=checkers,
+        grace_delay=config.getvalue("leaks_timeout"),
+        mark_failed=config.getvalue("leaks_fail"),
+    )
+    config.pluginmanager.register(checker, "leaks_checker")
+
 
+all_checkers: dict[str, type[ResourceChecker]] = {}
 
-all_checkers = {}
 
+class ResourceChecker:
+    name: ClassVar[str]
 
-def register_checker(name):
-    def decorate(cls):
-        assert issubclass(cls, ResourceChecker), cls
+    def __init_subclass__(cls, name: str):
         assert name not in all_checkers
+        cls.name = name
         all_checkers[name] = cls
-        return cls
-
-    return decorate
 
-
-class ResourceChecker:
-    def on_start_test(self):
+    def on_start_test(self) -> None:
         pass
 
-    def on_stop_test(self):
+    def on_stop_test(self) -> None:
         pass
 
-    def on_retry(self):
+    def on_retry(self) -> None:
         pass
 
-    def measure(self):
+    def measure(self) -> Any:
         raise NotImplementedError
 
-    def has_leak(self, before, after):
+    def has_leak(self, before: Any, after: Any) -> bool:
         raise NotImplementedError
 
-    def format(self, before, after):
+    def format(self, before: Any, after: Any) -> str:
         raise NotImplementedError
 
 
-@register_checker("fds")
-class FDChecker(ResourceChecker):
-    def measure(self):
-        if os.name == "posix":
-            import psutil
+class DemoChecker(ResourceChecker, name="demo"):
+    """Checker that always leaks. Used to test the core LeakChecker functionality."""
 
-            return psutil.Process().num_fds()
-        else:
+    i: int
+
+    def __init__(self):
+        self.i = 0
+
+    def measure(self) -> int:
+        self.i += 1
+        return self.i
+
+    def has_leak(self, before: int, after: int) -> bool:
+        return after > before
+
+    def format(self, before: int, after: int) -> str:
+        return f"Counter increased from {before} to {after}"
+
+
+class FDChecker(ResourceChecker, name="fds"):
+    def __init__(self):
+        BaseTCPConnector.warmup()
+
+    def measure(self) -> int:
+        if WINDOWS:
             return 0
```

Comment:
how about grabbing handles on Windows?

```suggestion
            return psutil.Process().num_handles()
```

Reply:
done

Reply:
Note that Windows handles encompass quite a bit more than files. For example Windows semaphores or mutexes are represented by handles as well. So you might get false positives here.

Reply:
Yes, the check on Windows is now a *lot* noisier than Linux and MacOS. It reports 21232 leaked handles by the end of it.
I confess my ignorance on low level Windows programming. Shouldn't semaphores and mutexes be cleaned up though? Is it normal/healthy to just leave them open like that?

Reply:
Reverting for the time being

Reply:
Semaphores and mutexes could be cached anywhere (for example in a third-party library). Also, keeping them allocated doesn't have the same consequences as keeping a file or socket open.